### DISCUSSION TITLE
Inquiry about File Management and Accessing Pyodide FS in Stlite

### DISCUSSION THREAD

**Discussion by kyh174** [whitphx/stlite#1108] [Q&A]
Hello,
I am a student from a university in Korea using Stlite for a project. I am creating a project with Google Blockly and Stlite.

This project allows users to code with blocks using Blockly, display the results with Stlite, and publish them as well (please refer to img01).
<img width="1512" alt="img01" src="https://github.com/user-attachments/assets/3a0b1ad7-3e94-451a-9928-370602f27a0d">

The goal of this project is to let users create their own machine learning models, similar to Google’s Teachable Machine, and combine them with Stlite to create interactive results, which can then be shared freely with others.

To the point:
I am trying to create a file tree for this project (please refer to img02).
<img width="1512" alt="img02" src="https://github.com/user-attachments/assets/4ad1618c-3dcd-4328-8661-a016ed873297">

I have implemented the handling of files uploaded to the file tree so they can be used inside Stlite’s Pyodide. However, I encountered a problem when trying to download the modified or preprocessed data back to the local environment, which led me to write this email.

In addition to this project, I am also working on another project where users can learn Python and machine learning by coding with blocks using Blockly and Pyodide on the web. This new project focuses on interactivity and publishing, which is why I am now using Stlite.

In the original project using Pyodide, I was able to create a file tree using Pyodide’s FS object, allowing me to upload files to Pyodide, download files from it, and modify files.

In the current project using Stlite, I don’t have access to the Pyodide object, so I have only been able to implement file uploading by using Stlite’s limited writeFile() method. However, I would like to implement file downloading as well, which is why I am reaching out for help.

Q1. Is there a way to access Pyodide’s FS object in Stlite?

Q2. If accessing FS is not possible, is there a way to handle downloading files from Pyodide to the local environment through the file tree, without writing Python code inside Stlite?

Lastly, I want to express my sincere gratitude as a user of Stlite working on this project.
(Please note that this email was written in Korean and translated into English using ChatGPT.)

**Comment by whitphx** [comment]
Hi, thank you for the message. I'm so glad to know such a passionate developer is using Stlite to create the wonderful works.

> Q1. Is there a way to access Pyodide’s FS object in Stlite?

No, because, in the case of Stlite, [Pyodide is working in a WebWorker](https://pyodide.org/en/stable/usage/webworker.html) so you can't access it directly from the JS main thread. This is also the reason why Stlite implemented its own methods such as `writeFile()`.

> Q2. If accessing FS is not possible, is there a way to handle downloading files from Pyodide to the local environment through the file tree, without writing Python code inside Stlite?

No, for now, unfortunately.
I created an issue to implement the methods to solve this problem: https://github.com/whitphx/stlite/issues/1110. Please track the progress on that issue.

Thanks!

**Reply by kyh174** [reply]
Hello,
Thank you so much for taking the time to kindly respond to my questions.
I will continue to work hard to further improve the project.
Once again, I sincerely appreciate your quick response and support.

**Reply by whitphx** [reply]
I prototyped the `readFile()` method in https://github.com/whitphx/stlite/pull/1125 and you can try it with the `mountable` build downloadable from https://github.com/whitphx/stlite/actions/runs/10977838242/artifacts/1962539122.

```javascript
const controller = stlite.mount(
/* ... options ... */
);

// After the app is launched...
const uint8Data = await controller.readFile(fileName);
```

Do you think this API satisfies your needs?

**Reply by kyh174** [reply]
Hello,
I apologize for the delayed response.
I tested the build you provided, and it works as I expected.
Thank you very much for your help.

Is there any possibility that the readFile() API will be officially added in the future?

Once again, I sincerely appreciate your help.
I will continue to work hard to improve my project. Thank you so much!

**Reply by whitphx** [reply]
Hi, thank you for checking it.
Yes I can merge it into the main release as users like you demand it.
One thing to note is that I'm not using `.readFile()` in my products such as Stlite Sharing so I can't maintain and keep checking it as frequently as other APIs such as `.writeFile()`.

**Reply by kyh174** [reply]
Thank you for the update and for considering the merge into the main release! I understand the focus on frequently used APIs like .writeFile() and appreciate the effort you're putting into it. Regarding the less frequent use of .readFile() in your products like Stlite Sharing, that makes sense. I'll keep that in mind moving forward.

Looking forward to the upcoming release!

**Reply by whitphx** [reply]
Thanks!
`.readFile()` is available in the latest release 0.68.0 👍
